File: backend/src/main/kotlin/net/animeta/backend/service/WorkService.kt
package net.animeta.backend.service

import net.animeta.backend.dto.Episode
import net.animeta.backend.model.TitleMapping
import net.animeta.backend.model.Work
import net.animeta.backend.repository.HistoryRepository
import net.animeta.backend.repository.TitleMappingRepository
import net.animeta.backend.repository.WorkRepository
import org.springframework.stereotype.Service

@Service
class WorkService(private val workRepository: WorkRepository,
                  private val titleMappingRepository: TitleMappingRepository,
                  private val historyRepository: HistoryRepository) {
    fun getOrCreate(title: String): Work {
        val mapping = titleMappingRepository.findOneByTitle(title)
        if (mapping != null) {
            return mapping.work
        }
        val key = normalizeTitle(title)
        val similarMapping = titleMappingRepository.findFirstByKey(key)
        if (similarMapping != null) {
            val mapping = titleMappingRepository.save(TitleMapping(
                    work = similarMapping.work,
                    title = title,
                    key = key
            ))
            return mapping.work
        } else {
            val work = workRepository.save(Work(
                    title = title,
                    image_filename = null,
                    raw_metadata = null,
                    metadata = null,
                    blacklisted = false
            ))
            titleMappingRepository.save(TitleMapping(
                    work = work,
                    title = title,
                    key = key
            ))
            return work
        }
    }

    fun getEpisodes(work: Work): List<Episode> {
        val result = historyRepository.findAllStatusWithCountAndCommentByWork(work)
                .mapNotNull { (status, count) ->
                    val statusNumber = status.toIntOrNull()
                    if (statusNumber != null)
                        Episode(number = statusNumber, post_count = count)
                    else
                        null
                }
                .associateBy { it.number }
                .toMutableMap()
        val result2 = historyRepository.findAllStatusWithCountAndNoCommentByWork(work)
        for ((status, count) in result2) {
            val statusNumber = status.toIntOrNull()
            if (statusNumber != null && !result.containsKey(statusNumber) && count > 1) {
                result[statusNumber] = Episode(number = statusNumber, post_count = null)
            }
        }
        return result.values.sortedBy { it.number }
    }

    companion object {
        private const val exceptionChars = "!+"

        fun normalizeTitle(title: String): String {
            return title
                    .map {
                        // full width -> half width
                        if (it in '\uFF01'..'\uFF5E') {
                            it.minus(0xFF01).plus(0x21)
                        } else {
                            it
                        }
                    }
                    .filter { exceptionChars.contains(it) || "LN".contains(it.category.code[0]) }
                    .joinToString("")
                    .toLowerCase()
                    .trim()
        }
    }
}

File: backend-ts/src/services/work.service.ts
import { Injectable } from "@nestjs/common";
import { InjectEntityManager, InjectRepository } from "@nestjs/typeorm";
import * as DataLoader from "dataloader";
import { Episode } from "shared/types_generated";
import { History } from "src/entities/history.entity";
import { TitleMapping } from "src/entities/title_mapping.entity";
import { Work } from "src/entities/work.entity";
import { WorkIndex } from "src/entities/work_index.entity";
import { objResults } from "src/utils/dataloader";
import { EntityManager, Repository } from "typeorm";
import { ValidationError } from "./exceptions";

@Injectable()
export class WorkService {
  private dataLoader = new DataLoader<number, Work>(
    objResults(ids => this.load(ids), k => `${k}`, v => `${v.id}`),
    { cache: false }
  );
  private indexDataLoader = new DataLoader<number, WorkIndex>(
    objResults(ids => this.loadIndex(ids), k => `${k}`, v => `${v.work_id}`),
    { cache: false }
  );
  
  constructor(
    @InjectRepository(Work) private workRepository: Repository<Work>,
    @InjectRepository(WorkIndex) private workIndexRepository: Repository<WorkIndex>,
    @InjectRepository(History) private historyRepository: Repository<History>,
    @InjectEntityManager() private entityManager: EntityManager,
  ) {}

  get(id: number): Promise<Work> {
    return this.dataLoader.load(id);
  }

  getIndex(id: number): Promise<WorkIndex> {
    return this.indexDataLoader.load(id);
  }

  async getEpisodes(work: Work): Promise<Episode[]> {
    const result = await this.historyRepository.createQueryBuilder()
      .select('status')
      .addSelect('COUNT(*)', 'count')
      .where('work_id = :workId AND comment <> :commentNot', { workId: work.id, commentNot: '' })
      .groupBy('status')
      .getRawMany()
    const result2 = await this.historyRepository.createQueryBuilder()
      .select('status')
      .addSelect('COUNT(*)', 'count')
      .where('work_id = :workId AND comment = :comment', { workId: work.id, comment: '' })
      .groupBy('status')
      .getRawMany()
    const episodeMap = new Map<number, Episode>()
    for (const { status, count } of result) {
      if (!/^[0-9]+$/.test(status)) continue
      const statusNumber = Number(status)
      episodeMap.set(statusNumber, {
        number: statusNumber,
        post_count: Number(count),
      })
    }
    for (const { status, count } of result2) {
      if (!/^[0-9]+$/.test(status)) continue
      const statusNumber = Number(status)
      const countNumber = Number(count)
      if (!episodeMap.has(statusNumber) && countNumber > 1) {
        episodeMap.set(statusNumber, {
          number: statusNumber,
          post_count: null,
        })
      }
    }
    return Array.from(episodeMap.values()).sort((a, b) => a.number - b.number)
  }

  private async load(ids: readonly number[]): Promise<Work[]> {
    return this.workRepository.findByIds(Array.from(ids));
  }

  private async loadIndex(ids: readonly number[]): Promise<WorkIndex[]> {
    return this.workIndexRepository.findByIds(Array.from(ids));
  }

  getImageUrl(work: Work): string | null {
    return getWorkImageUrl(work)
  }

  async getOrCreate(title: string): Promise<Work> {
    return getOrCreateWork(this.entityManager, title)
  }
}

export function getWorkImageUrl(work: Work): string | null {
  // TODO: config
  return work.image_filename ? `https://storage.googleapis.com/animeta-static/media/${work.image_filename}` : null
}

export async function getOrCreateWork(em: EntityManager, title: string): Promise<Work> {
  title = title.trim()
  if (title === '')
    throw new ValidationError('작품 제목을 입력하세요.')
  
  const mapping = await em.findOne(TitleMapping, { where: {title} })
  if (mapping) {
    return em.findOneOrFail(Work, mapping.work_id)
  }
  const key = normalizeTitle(title)
  const similarMapping = await em.findOne(TitleMapping, { where: {key} })
  if (similarMapping) {
    await em.save(TitleMapping, {
      work_id: similarMapping.work_id,
      title,
      key
    })
    return em.findOneOrFail(Work, similarMapping.work_id)
  }
  const work = new Work()
  work.title = title
  work.image_filename = null
  work.original_image_filename = null
  work.image_center_y = 0.0
  work.raw_metadata = null
  work.metadata = null
  work.blacklisted = false
  work.first_period = null
  await em.save(work)
  await em.save(TitleMapping, {
    work_id: work.id,
    title,
    key,
  })
  return work
}

const exceptionChars = ['!', '+']

export function normalizeTitle(title: string): string {
  return Array.from(title)
    .map(c => {
      // full width -> half width
      if ('\uFF01' <= c && c <= '\uFF5E')
        return String.fromCodePoint(c.codePointAt(0)! - 0xFF01 + 0x21)
      else
        return c
    })
    .filter(c => exceptionChars.includes(c) || /\p{L}|\p{N}/u.test(c))
    .join('')
    .toLowerCase()
    .trim()
}
